### PR TITLE
feat: plex integration

### PR DESCRIPTION
**Pull request by hillaliy**
<br/>
<div align="center">
  
  <h3>Homarr</h3>
</div>

**Thank you for your contribution. Please ensure that your pull request meets the following pull request:**

- [x] Builds without warnings or errors (``pnpm buid``, autofix with ``pnpm format:fix``)
- [x] Pull request targets ``dev`` branch
- [x] Commits follow the [conventional commits guideline](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] No shorthand variable names are used (eg. ``x``, ``y``, ``i`` or any abbrevation)

### CODE REVIEW

**Comment by hillaliy** [comment]
I need some help to fix lint error

**Comment by manuel-rw** [comment]
You should move ``@types/xml2js`` to ``devDependencies`` because it's never needed in the production. It's only used when TypeScript builds / lints the app.

**Comment by hillaliy** [comment]
sorry for many commits

File: packages/integrations/src/plex/plex-integration.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    const token = super.getSecretValue("apiKey");
+
+    const response = await fetch(`${this.integration.url}/status/sessions?X-Plex-Token=${token}`);
+    const body = await response.text();
+
+    // convert xml response to objects, as there is no JSON api
+    const data = (await parseStringPromise(body)) as PlexResponse;
+    const mediaContainer = data.MediaContainer;
+    // no sessions are open or available
+    if (!mediaContainer.Video) {
```
Comment (manuel-rw) [inline packages/integrations/src/plex/plex-integration.ts:19]: debug log here

**Review by manuel-rw — CHANGES_REQUESTED** [review]

File: packages/integrations/src/plex/plex-integration.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    const token = super.getSecretValue("apiKey");
+
+    await super.handleTestConnectionResponseAsync({
+      queryFunctionAsync: async () => {
+        return await fetch(`${this.integration.url}?X-Plex-Token=${token}`);
+      },
+      handleResponseAsync: async (response) => {
+        const result = await response.text();
+        const parsedResponse = (await parseStringPromise(result)) as unknown;
+        if (typeof parsedResponse === "object" && parsedResponse !== null) {
```
Comment (SeDemal) [inline packages/integrations/src/plex/plex-integration.ts:77]: You should check for specific data, any URL with an xml response would return valid.
Just checking for the existence of the "MediaContainer" element should be enough.

File: packages/integrations/src/plex/plex-integration.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        return {
+          sessionId: sessionElement?.$.id ?? "unknown",
+          sessionName: `${playerElement.$.product} (${playerElement.$.title})`,
+          user: {
+            userId: userElement?.$.id ?? "Anonymous",
+            username: userElement?.$.title ?? "Anonymous",
+            profilePictureUrl: userElement?.$.thumb ?? null,
+          },
+          currentlyPlaying: {
+            type: PlexIntegration.getCurrentlyPlayingType(videoElement.$.type) ?? "video",
```
Comment (SeDemal) [inline packages/integrations/src/plex/plex-integration.ts:49]:
```TS
type: videoElement.$.live === "1" ? "tv" : this.getCurrentlyPlayingType(videoElement.$.type),
```

**Review by SeDemal — CHANGES_REQUESTED** [review]
I haven't been able to try it much, but I found this: https://plexapi.dev/Intro
I'm not sure if it actually is native plex or for a plugin, but they seem to handle the responses as JSON, and they provide an sdk.
The xml2js option is good enough, I'm just saying it's worth looking into.

File: packages/integrations/src/plex/plex-integration.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import { Integration } from "../base/integration";
+import { IntegrationTestConnectionError } from "../base/test-connection-error";
+import type { StreamSession } from "../interfaces/media-server/session";
+import type { PlexResponse } from "./interface";
+
+export class PlexIntegration extends Integration {
+  public async getCurrentSessionsAsync(): Promise<StreamSession[]> {
+    const token = super.getSecretValue("apiKey");
+
+    const response = await fetch(`${this.integration.url}/status/sessions?X-Plex-Token=${token}`);
```
[the remark this answers is not in the document]
Comment (SeDemal) [reply packages/integrations/src/plex/plex-integration.ts]: You fixed it already when I was doing my review, nvm good job

File: packages/integrations/src/plex/plex-integration.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    return videos;
+  }
+
+  public async testConnectionAsync(): Promise<void> {
+    const token = super.getSecretValue("apiKey");
+
+    await super.handleTestConnectionResponseAsync({
+      queryFunctionAsync: async () => {
+        return await fetch(`${this.integration.url}?X-Plex-Token=${token}`);
```
[the remark this answers is not in the document]
Comment (SeDemal) [reply packages/integrations/src/plex/plex-integration.ts]: You fixed it already when I was doing my review, nvm good job

**Review by manuel-rw — DISMISSED** [review]
Please wait for other reviewers to approve

File: packages/integrations/src/plex/plex-integration.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    const response = await fetch(`${this.integration.url}/status/sessions`, {
+      headers: {
+        "X-Plex-Token": token,
+      },
+    });
+    const body = await response.text();
+    // convert xml response to objects, as there is no JSON api
+    const data = await PlexIntegration.parseXml<PlexResponse>(body);
+    const mediaContainer = data.MediaContainer;
```
Comment (SeDemal) [inline packages/integrations/src/plex/plex-integration.ts:22]: Funny thing, oldmarr used element[0], meaning it didn't actually target a "MediaContainer".
In the case of music, the element is actually a "Track" and not a "MediaContainer".
Oldmarr was working by mistake. In your case though, you need to handle "data.Track" instead for music.

**Review by SeDemal — CHANGES_REQUESTED** [review]

**Pull request merged**